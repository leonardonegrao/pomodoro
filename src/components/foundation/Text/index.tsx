import styled, { css } from 'styled-components';

type TextVariant = 'appTitle' | 'heading1' | 'heading2' | 'heading3' | 'heading4' | 'body1' | 'body2';

export const TextVariantsMap = (variant: TextVariant) => css`
  font-size: ${({ theme }) => theme.typography.variants[variant].fontSize};
  font-weight: ${({ theme }) => theme.typography.variants[variant].fontWeight};
  line-height: ${({ theme }) => theme.typography.variants[variant].lineHeight};

  ${({ theme }) => {
    if (theme.typography.variants[variant].letterSpacing) {
      return css`
        letter-spacing: ${theme.typography.variants[variant].letterSpacing};
    `;
    }

    if (theme.typography.variants[variant].textTransform) {
      return css`
        text-transform: ${theme.typography.variants[variant].textTransform};
    `;
    }

    return css``;
  }}
`;

interface BaseTypographyProps {
  variant: TextVariant
  as?: React.ElementType
}

const BaseTypography = styled.span<BaseTypographyProps>`
  ${({ variant }) => TextVariantsMap(variant)}
  color: ${({ theme }) => theme.colors.white};
  margin: 0;
`;

interface TextProps extends React.HTMLAttributes<HTMLSpanElement> {
  tag?: React.ElementType;
  variant: TextVariant;
  children: React.ReactNode;
  color: string;
}

export default function Text({
  tag = 'span', color, variant, children, ...props
}: TextProps) {
  return (
    <BaseTypography
      as={tag}
      color={color}
      variant={variant}
      {...props}
    >
      {children}
    </BaseTypography>
  );
}
