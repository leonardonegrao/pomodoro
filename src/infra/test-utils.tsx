import { render, RenderOptions } from '@testing-library/react';
import { FC, ReactElement } from 'react';

import { ThemeProvider } from '@shared/contexts/ThemeProvider';
import { TimersProvider } from '@shared/contexts/TimersProvider';


const AllTheProviders: FC = ({ children }) => (
  <ThemeProvider>
    <TimersProvider>
      {children}
    </TimersProvider>
  </ThemeProvider>
);

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>,
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react';
export { customRender as render };
