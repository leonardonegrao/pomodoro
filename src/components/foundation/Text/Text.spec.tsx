import { render, screen } from '../../../infra/test-utils';
import Text from './index';

describe('<Text />', () => {
  it('should render component', () => {
    render(
      <Text
        color="white"
        variant="appTitle"
      >
        foo
      </Text>,
    );

    const text = screen.getByText('foo');
    expect(text).toMatchSnapshot();
  });

  it('should render as the HTML element passed by props', () => {
    render(
      <Text
        as="h3"
        color="white"
        variant="heading3"
      >
        foo
      </Text>,
    );

    const text = screen.getByText('foo');

    expect(text.nodeName).toBe('H3');
  });

  it('should render using the color passed by props', () => {
    const { container } = render(
      <Text
        as="h3"
        color="primary.main"
        variant="heading3"
      >
        foo
      </Text>,
    );

    expect(container.firstChild).toHaveStyle('color: #1E213F');
  });

  it('should apply any options passed by props', () => {
    render(
      <Text
        as="h3"
        color="primary.main"
        options={{
          opacity: 0.2,
          fontWeight: '500',
          lineHeight: '30px',
        }}
        variant="heading3"
      >
        foo
      </Text>,
    );

    const text = screen.getByText('foo');

    expect(text).toHaveStyle('opacity: 0.2');
    expect(text).toHaveStyle('font-weight: 500');
    expect(text).toHaveStyle('line-height: 30px');
  });
});
