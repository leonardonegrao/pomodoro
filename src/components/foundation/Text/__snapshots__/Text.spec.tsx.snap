// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Text /> should render component 1`] = `
.c0 {
  color: #FFFFFF;
  font-family: "Kumbh Sans",sans-serif;
  font-size: 24px;
  font-weight: bold;
  line-height: 24.14px;
  margin: 0;
}

@media (min-width:768px) {
  .c0 {
    font-size: 32px;
    line-height: 32.19px;
  }
}

<span
  class="c0"
  color="white"
>
  foo
</span>
`;
